core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -363,6 +386,22 @@ class BrokerToControllerRequestThread(
     }
   }
 
+  def maybeCheckNotControllerErrorInsideEnvelopeResponse(requestHeader: RequestHeader, responseBody: AbstractResponse): Boolean = {
```

Comment:
How about move this to `ForwardingManager` since the envelopResponse and envelopRequest are all in `ForwardingManager`

Reply:
Yes, I tried, but it can't. The `activeController` is stored in `BrokerToControllerRequestThread`, which owned by `BrokerToControllerChannelManager`. So, only the `BrokerToControllerRequestThread#handleResponse` can update the activeController and put the reqeust into top of the queue again (retry).